src/main/java/org/hydev/mcpm/client/installer/DatabaseManager.java

```diff
@@ -0,0 +1,82 @@
+import org.hydev.mcpm.client.database.fetcher.LocalDatabaseFetcher;
+import org.hydev.mcpm.client.database.inputs.SearchPackagesInput;
+import org.hydev.mcpm.client.database.inputs.SearchPackagesType;
+import org.hydev.mcpm.client.database.results.ListPackagesResult;
+import org.hydev.mcpm.client.database.results.SearchPackagesResult;
+import org.hydev.mcpm.client.injector.PluginNotFoundException;
+import org.hydev.mcpm.client.installer.input.InstallInput;
+import org.hydev.mcpm.client.models.PluginModel;
+import org.hydev.mcpm.client.models.PluginVersion;
+import org.hydev.mcpm.client.models.PluginYml;
+
+import java.io.File;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+
+/*
+* An interactor that communicates with database and returns the plugin source for installation
+* */
+public class DatabaseManager {
+    private final DatabaseInteractor databaseInteractor;
+    private final LocalPluginTracker localPluginTracker;
+
+    /**
+     * Initialize an database manager
+     */
+    public DatabaseManager() {
+        this.databaseInteractor = new DatabaseInteractor(new LocalDatabaseFetcher(URI.create("http://mcpm.hydev.org")));
+        this.localPluginTracker = new LocalPluginTracker("lock", "plugins");
+    }
+
+    /**
+     * Search if the plugin exists, throw exception if it does not exist and return the information of
+     * the pluigin needs to be installed.
+     * @param installInput: the plugin input pending for installation
+     */
+    public SearchPackagesResult getSearchResult(InstallInput installInput) {
+        SearchPackagesInput searchPackagesInput = new SearchPackagesInput(installInput.type(), installInput.name(), false);
+        SearchPackagesResult searchPackageResult = databaseInteractor.search(searchPackagesInput);
+        switch (searchPackageResult.state()) {
+            case SUCCESS -> {
+                return searchPackageResult;
+            }
+            case INVALID_INPUT -> {
+                throw new IllegalArgumentException("Invalid Input");
+            }
+            case FAILED_TO_FETCH_DATABASE -> {
+                throw new RuntimeException("Fail to fetch database");
+            }
+        }
+        return searchPackageResult;
+    }
+
+    /**
+    * check if plugin already installed locally
+    * @param pluginVersion: The version of the installed plugin
+    * */
+    public void checkPluginInstalled(PluginVersion pluginVersion) {
+        List<PluginYml> pluginInstalled = localPluginTracker.listInstalled();
+        for (PluginYml pluginYml : pluginInstalled) {
+            if (pluginYml != null && pluginYml.name() != null && pluginYml.name().equals(pluginVersion.meta().name())) {
+                throw new RuntimeException();
+            }
+        }
+    }
+
+    /**
+    * Add the installed plugin to the csv file
+    * @param pluginVersion: The version of the installed plugin
+    * */
+    public void addManualInstalled(PluginVersion pluginVersion) {
```

Comment:
This does indeed seem like a use case method. If I understand correctly, this should be all for the installing use case. Why the need for two intearctors? That is, why not put this in installinteractor? You'll also need to create another boundary for this is an interactor.

Reply:
You have a good point. I was trying to let the installer interactor invokes the install method to serve its single-use case. If we add this add manual install, then it has to couple with more classes in the database and carry more responsibility by working on with database. So I just create another hook for the interactor (database manager) to check and prepare the input for the interactor. Another thing I found is that what if uninstall and update have to access to the database to carry out the same check like install. Then, I think a good choice is to separate them, and yes I would need an interface for the database manager if I do so. Feel free to correct my thought :D